core/src/subscription/quic_stream.rs

```diff
@@ -0,0 +1,156 @@
+    #[inline]
+    pub fn server(&self) -> SocketAddr {
+        self.five_tuple.resp
+    }
+}
+
+impl Subscribable for QuicStream {
+    type Tracked = TrackedQuic;
+
+    fn level() -> Level {
+        Level::Session
+    }
+
+    fn parsers() -> Vec<ConnParser> {
+        vec![ConnParser::Quic(QuicParser::default())]
+    }
+
+    fn process_packet(
+        mbuf: Mbuf,
+        subscription: &Subscription<Self>,
+        conn_tracker: &mut ConnTracker<Self::Tracked>,
+    ) {
+        match subscription.filter_packet(&mbuf) {
+            FilterResult::MatchTerminal(idx) | FilterResult::MatchNonTerminal(idx) => {
+                if let Ok(ctxt) = L4Context::new(&mbuf, idx) {
+                    conn_tracker.process(mbuf, ctxt, subscription);
+                }
+            }
+            FilterResult::NoMatch => drop(mbuf),
+        }
+    }
+}
+
+/// Represents a QUIC connection during the connection lifetime.
+///
+/// ## Remarks
+/// Retina uses an internal parser to track and filter application-layer protocols, and transfers
+/// session ownership to the subscription to invoke the callback on a filter match. This is an
+/// optimization to avoid double-parsing: once for the filter and once for the subscription data.
+/// This is why most `Trackable` trait methods for this type are unimplemented.
+///
+/// ## Note
+/// Internal connection state is an associated type of a `pub` trait, and therefore must also be
+/// public. Documentation is hidden by default to avoid confusing users.
+#[doc(hidden)]
+pub struct TrackedQuic {
+    five_tuple: FiveTuple,
+    connection_id: HashSet<String>,
+}
+
+impl TrackedQuic {
+    fn get_connection_id(&self, dcid_bytes: &[u8]) -> Option<String> {
+        let dcid_hex = Quic::vec_u8_to_hex_string(dcid_bytes);
+        for dcid in &self.connection_id {
+            if dcid_hex.starts_with(dcid) {
+                return Some(dcid.clone());
+            }
+        }
+        None
+    }
+}
+
+impl Trackable for TrackedQuic {
+    type Subscribed = QuicStream;
+
+    fn new(five_tuple: FiveTuple) -> Self {
+        TrackedQuic {
+            five_tuple,
+            connection_id: HashSet::new(),
+        }
+    }
+
+    fn pre_match(&mut self, _pdu: L4Pdu, _session_id: Option<usize>) {}
+
+    fn on_match(&mut self, session: Session, subscription: &Subscription<Self::Subscribed>) {
+        if let SessionData::Quic(quic) = session.data {
+            let mut quic_clone = (*quic).clone();
+
+            if let Some(long_header) = &quic_clone.long_header {
+                if long_header.dcid_len > 0 {
+                    self.connection_id.insert(long_header.dcid.clone());
+                }
+                if long_header.scid_len > 0 {
+                    self.connection_id.insert(long_header.scid.clone());
+                }
+            } else {
+                if let Some(ref mut short_header_value) = quic_clone.short_header {
+                    short_header_value.dcid =
```

Comment:
Based on only delivering short headers if connection id is present from long header, should you only deliver if get_connection_id returns None?

Reply:
Good question! No, because I only return a ProbeResult::Certain when the header is a long header because those give enough information to be certain. From my understanding, this then results in the connection becoming tracked so that future packets that match this 5 tuple also use the parse pathway, which means they also call the on_match function if successful in initial parsing. Given that it is a part of an existing 5 tuple I'm certain enough that it is a quic header, but there's still a chance that I cannot match the dcid from that packet and so no matter if I get None or Some in the response I'll still want to populate the dcid appropriately.

Reply:
Renamed Quic -> QuicPacket. Thank you for the insight!